### PR TITLE
Added Gmsh 4.1 file format parser

### PR DESCRIPTION
**Pull request by masonamccallum**
Created readGmsh2D_v4 parsing function for Gmsh version 4.1. I can now group elements into regions/physical groups in Gmsh and then maintain those groupings on import. This way Physical parameters can be added for processing. For example in the mesh below I wanted a parameter m to be different on either side of the middle interface. 

image

### CODE REVIEW

**Comment by jlchan** [comment]
@masonamccallum thanks very much for adding this functionality! 

I'll have a more detailed review tonight, but would you mind adding a test of your new Gmsh reader to this PR?

**Comment by jlchan** [comment]
If you'd like, a test using the mesh you constructed would be a cool setup.

File: src/mesh/simple_meshes.jl [hunk 1]
```diff
@@ -1,23 +1,212 @@
+function findline(name, lines)
+    for (i, line) in enumerate(lines)
+        if line == name
+            return i
+        end
+    end
+end
+
+struct MeshOptions
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:9]: Could you add a docstring describing what `MeshOptions` is used for and how it should be used?

File: src/mesh/simple_meshes.jl [hunk 2]
```diff
@@ -1,23 +1,212 @@
...
+    grouping::Bool
+    remapGroupName::Bool
+end
+
+"""
+ returns the number of elements in a .msh file of a specified dimension
+"""
+function getNumElements(lines,Dim=2)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:17]: Would you mind switching this to snake case `get_num_elements`? I'm trying to slowly bring the package more in line with the Julia style guide https://docs.julialang.org/en/v1/manual/style-guide/...

File: src/mesh/simple_meshes.jl [hunk 3]
```diff
@@ -1,23 +1,212 @@
...
+    elem_start = findline("\$Elements",lines) + 1
+    temp_numBlocks,_,_,_ = split(lines[elem_start])
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:19]:
```suggestion
    temp_numBlocks, _ = split(lines[elem_start])
```
I think the trailing `_` just discards other output arguments.

File: src/mesh/simple_meshes.jl [hunk 4]
```diff
@@ -1,23 +1,212 @@
...
+    numBlocks = parse(Int,temp_numBlocks)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:20]:
```suggestion
    numBlocks = parse(Int, temp_numBlocks)
```

File: src/mesh/simple_meshes.jl [hunk 5]
```diff
@@ -1,23 +1,212 @@
...
+    numElements = 0
+    block_data_line = elem_start + 1
+    for i in 1:numBlocks
+        temp_blockDim,_,_,temp_num_elem_in_block = split(lines[block_data_line])
+        blockDim = parse(Int,temp_blockDim)
+        num_elem_in_block = parse(Int,temp_num_elem_in_block)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:26]:
```suggestion
        temp_blockDim, _, _, temp_num_elem_in_block = split(lines[block_data_line])
        blockDim = parse(Int,t emp_blockDim)
        num_elem_in_block = parse(Int, temp_num_elem_in_block)
```

File: src/mesh/simple_meshes.jl [hunk 6]
```diff
@@ -1,23 +1,212 @@
...
+        end
+        block_data_line = block_data_line + num_elem_in_block + 1
+    end
+    return numElements
+end
+"""
+GMSH uses integers for naming conventions. This function remaps the gmsh ids to
+a list of ids 1:numGroups. This just cleans up a little after Gmsh
+"""
+function remapElementGrouping!(eg)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:38]:
```suggestion
function remap_element_grouping!(eg)
```
see above note on the Julia style guide

File: src/mesh/simple_meshes.jl [hunk 7]
```diff
@@ -1,23 +1,212 @@
...
+option = MeshOption(true)
+VXY, EToV, grouping = readGmsh2D_v4("eulerSquareCylinder2D.msh", option)
+```
+https://gmsh.info/doc/texinfo/gmsh.html#MSH-file-format
+
+Notes:The version 4 format has a more detailed block data format
+this leads to more complicated parser.
+"""
+function readGmsh2D_v4(filename::String, options::MeshOptions)
+    @unpack grouping,remapGroupName = options;
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:76]:
```suggestion
    @unpack grouping, remapGroupName = options;
```

File: src/mesh/simple_meshes.jl [hunk 8]
```diff
@@ -1,23 +1,212 @@
...
+    f = open(filename)
+    lines = readlines(f)
+
+    format_line = findline("\$MeshFormat",lines)+1
+    version,_,dataSize = split(lines[format_line])
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:81]:
```suggestion
    version, _, dataSize = split(lines[format_line])
```

File: src/mesh/simple_meshes.jl [hunk 9]
```diff
@@ -1,23 +1,212 @@
...
+    gmsh_version = parse(Float64,version)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:82]:
```suggestion
    gmsh_version = parse(Float64, version)
```

File: src/mesh/simple_meshes.jl [hunk 10]
```diff
@@ -1,23 +1,212 @@
...
+    end
+
+    if grouping
+        # For 2D Physical group data we are only interested in groupings in surface entities
+        # This will extract relevent grouping data for each surface
+        # only supports one physical group per Element
+        entities_start_line = findline("\$Entities", lines) +1
+        entities_block_data = split(lines[entities_start_line])
+        entities_block_data = [parse(Int,c) for c in entities_block_data[1:3]]
+        numPoints,numCurves,numSurfaces = entities_block_data[1:3]
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:97]:
```suggestion
        numPoints, numCurves, numSurfaces = entities_block_data[1:3]
```

File: src/mesh/simple_meshes.jl [hunk 11]
```diff
@@ -1,23 +1,212 @@
...
+            surfaceData[surfTag] = PhysTag
+        end
+        @info "Tag grouping: $surfaceData"
+    end
+
+    node_start = findline("\$Nodes", lines) + 1
+    temp_numBlocks,temp_Nv,_,_ = split(lines[node_start])
+    numBlocks = parse(Int,temp_numBlocks)
+    Nv = parse(Int,temp_Nv)
+    VX,VY,VZ = ntuple(x->zeros(Float64,Nv),3)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:116]:
```suggestion
    temp_numBlocks, temp_Nv, _ = split(lines[node_start])
    numBlocks = parse(Int, temp_numBlocks)
    Nv = parse(Int, temp_Nv)
    VX, VY, VZ = ntuple(x->zeros(Float64, Nv), 3)
```

File: src/mesh/simple_meshes.jl [hunk 12]
```diff
@@ -1,23 +1,212 @@
...
+    block_line = node_start + 1
+    for block_idx in 1:numBlocks
+        _,_,_,temp_numNodes = split(lines[block_line])
+        numNodes = parse(Int,temp_numNodes)
+        for block_node_idx in 1:numNodes
+            node_num_line = block_line + block_node_idx 
+            node_cords_line = node_num_line + numNodes 
+            node_idx_global = parse(Int,lines[node_num_line])
+            vals = [parse(Float64,c) for c in split(lines[node_cords_line])]
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:125]:
```suggestion
        _, _, _, temp_numNodes = split(lines[block_line])
        numNodes = parse(Int, temp_numNodes)
        for block_node_idx in 1:numNodes
            node_num_line = block_line + block_node_idx 
            node_cords_line = node_num_line + numNodes 
            node_idx_global = parse(Int, lines[node_num_line])
            vals = [parse(Float64, c) for c in split(lines[node_cords_line])]
```

File: src/mesh/simple_meshes.jl [hunk 13]
```diff
@@ -1,23 +1,212 @@
...
+            VX[node_idx_global] = vals[1]
+            VY[node_idx_global] = vals[2]
+        end
+        block_line = block_line + 2*numNodes+ 1
+    end
+
+    elem_start = findline("\$Elements",lines) + 1
+    temp_numBlocks,_,_,_ = split(lines[elem_start])
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:133]:
```suggestion
    temp_numBlocks, _ = split(lines[elem_start])
```

File: src/mesh/simple_meshes.jl [hunk 14]
```diff
@@ -1,23 +1,212 @@
...
+    numElements = getNumElements(lines) # Some elements in the .msh file are not our 2d mesh elements see docstring of function
+    @info "File has $numElements 2D Elements"
+    numBlocks = parse(Int, temp_numBlocks)
+
+    if grouping
+        element_grouping = zeros(Int,numElements) #element physical group assignment
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:139]:
```suggestion
        element_grouping = zeros(Int, numElements) #element physical group assignment
```

File: src/mesh/simple_meshes.jl [hunk 15]
```diff
@@ -1,23 +1,212 @@
...
+    end
+
+    EToV = zeros(Int64,numElements,3)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:142]:
```suggestion
    EToV = zeros(Int64, numElements, 3)
```

File: src/mesh/simple_meshes.jl [hunk 16]
```diff
@@ -1,23 +1,212 @@
...
+    block_line_start = elem_start + 1
+    for block in 1:numBlocks 
+        temp_elem_block_dim,temp_tag,_,temp_elemInBlock = split(lines[block_line_start])
+        elem_block_dim = parse(Int, temp_elem_block_dim)
+        numElemInBlock = parse(Int,temp_elemInBlock)
+        surface_tag = parse(Int,temp_tag)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:148]:
```suggestion
        temp_elem_block_dim, temp_tag, _, temp_elemInBlock = split(lines[block_line_start])
        elem_block_dim = parse(Int, temp_elem_block_dim)
        numElemInBlock = parse(Int, temp_elemInBlock)
        surface_tag = parse(Int, temp_tag)
```

File: src/mesh/simple_meshes.jl [hunk 17]
```diff
@@ -1,23 +1,212 @@
...
+        if elem_block_dim == 2 # only interesed in 2d triangle elements
+            for e_idx in 1:numElemInBlock
+                vals = [parse(Int,c) for c in split(lines[e_idx+block_line_start])]
+                elem_global_idx, nodeA, nodeB, nodeC = vals
+                EToV[elem_global_idx,:] .= [nodeA,nodeB,nodeC]
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:153]:
```suggestion
                EToV[elem_global_idx,:] .= [nodeA, nodeB, nodeC]
```

File: src/mesh/simple_meshes.jl [hunk 18]
```diff
@@ -1,23 +1,212 @@
...
 
-    function findline(name, lines)
-        for (i, line) in enumerate(lines)
-            if line == name
-                return i
-            end
-        end
+    format_line = findline("\$MeshFormat",lines)+1
+    version,_,dataSize = split(lines[format_line])
+    gmsh_version = parse(Float64,version)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:203]:
```suggestion
    format_line = findline("\$MeshFormat", lines)+1
    version, _, dataSize = split(lines[format_line])
    gmsh_version = parse(Float64, version)
```

**Review by jlchan — CHANGES_REQUESTED** [review]
Thanks for adding this functionality! 

Here are some suggestions - feel free to ignore or discard any of them. I mostly added spacing between commas (input/output arguments) and some formatting. 

Once we get this merged, I'll release an update. I'd eventually like to make it so that the version is just automatically detected in the future.

**Comment by jlchan** [comment]
Feel free to ping me if you have any questions or would like me to take another look!

File: src/mesh/simple_meshes.jl [hunk 19]
```diff
@@ -1,23 +1,233 @@
...
+"""
+function getNumElements(lines,Dim=2)
+    elem_start = findline("\$Elements",lines) + 1
+    temp_numBlocks, _ = split(lines[elem_start])
+    numBlocks = parse(Int, temp_numBlocks)
+    numElements = 0
+    block_data_line = elem_start + 1
+    for i in 1:numBlocks
+        temp_blockDim, _, _, temp_num_elem_in_block = split(lines[block_data_line])
+        blockDim = parse(Int,t emp_blockDim)
```
Comment (jlchan) [inline src/mesh/simple_meshes.jl:25]:
```suggestion
        blockDim = parse(Int, temp_blockDim)
```

Comment (jlchan) [reply src/mesh/simple_meshes.jl]: Fixing a bug I introduced

**Comment by masonamccallum** [comment]
Okay great! Ive made a bunch of changes to clean things up. I should be able to commit those fixes and the new tests early tomorrow. Thanks for the feedback!

**Comment by jlchan** [comment]
@masonamccallum it looks like the tests for Julia v1.6 are failing. Can you take a look at those? If you need I can also try to figure out what's going on.

**Comment by masonamccallum** [comment]
Ahh I used a feature in the testing library that requires Julia 1.7.

**Comment by jlchan** [comment]
It looks like the tests are passing now. If you're cool with this PR as is, I can merge it and release a new version after they've passed.

**Comment by masonamccallum** [comment]
Yeah that would be great! I was planning on setting up a function read_gmsh_2d that would determine the file format. I wanted your option on what it should return. The following would be a straight forward implementation.

`(VY,VX),EToV = read_gmsh_2d(filename)`
 
I would however like to get other data from the .msh files beyond just VXY and EToV. The file format for Gmsh 4.1 however doesn't lend itself well to implementing a separate function for extracting additional data. I think it would be better implementation to return a named tuple (or custom type) with relevant gmsh data that can then be unpacked.

**Comment by jlchan** [comment]
> Yeah that would be great! I was planning on setting up a function read_gmsh_2d that would determine the file format automatically. I wanted your option on what it should return. The following would be a straight forward implementation.
> 
> `(VY,VX),EToV = read_gmsh_2d(filename)`

Minor note - should it be `(VX, VY)` instead of `(VY, VX)`?
> 
> I would however like to get other data from the .msh files beyond just VXY and EToV. The file format for Gmsh 4.1 however doesn't lend itself well to implementing a separate function for extracting additional data. I think it would be better implementation to return a named tuple (or custom type) with relevant gmsh data that can then be unpacked.

This sounds great. I think you could simply have the additional data returned at the end - the output of `read_gmsh_2d` is a tuple, and if you have more output arguments than entries of this tuple, then it will just extract the first few.

**Comment by jlchan** [comment]
Another alternative to unifying the output arguments would be to build a custom struct for GMSH outputs and customize `iterate`. This does have the advantage of allowing dispatch on the mesh if we wanted to do so in the future.

**Comment by jlchan** [comment]
Is the new implementation something you'd want to include in this PR or a later one?

**Comment by masonamccallum** [comment]
A later PR is fine. At the moment I think the functionality is well tested and working.

**Comment by jlchan** [comment]
Great, then I'll go ahead and merge this. Can you make an issue describing the new implementation?

**Comment by masonamccallum** [comment]
Of course. Im on it

**Comment by jlchan** [comment]
Registration for new version 0.13.8 at https://github.com/JuliaRegistries/General/pull/65213

**Comment by jlchan** [comment]
And new version - https://github.com/jlchan/StartUpDG.jl/releases/tag/v0.13.8. Thanks again for this feature, @masonamccallum!

**Pull request merged**